product_profile/product.py

```diff
@@ -80,56 +80,48 @@ def _get_profile_fields(self):
 
     @api.model
     def _get_profile_data(self, profile_id, filled_fields=None):
+        # Note for migration to v9
+        # - rename method to a more convenient name _get_vals_from_profile()
+        # - remove unused filled_fields args
         profile_obj = self.env['product.profile']
         fields = self._get_profile_fields()
-        if profile_id:
-            profile = profile_obj.browse(profile_id).read(fields)[0]
-            profile.pop('id')
-            for field, value in profile.items():
-                if value and profile_obj._fields[field].type == 'many2one':
-                    # m2o value is a tuple
-                    profile[field] = value[0]
-                if profile_obj._fields[field].type == 'many2many':
-                    profile[field] = [(6, 0, value)]
-            return profile
-        else:
-            return {field: None for field in fields}
+        vals = profile_obj.browse(profile_id).read(fields)[0]
+        vals.pop('id')
+        for field, value in vals.items():
+            if value and profile_obj._fields[field].type == 'many2one':
+                # m2o value is a tuple
+                vals[field] = value[0]
+            if profile_obj._fields[field].type == 'many2many':
+                vals[field] = [(6, 0, value)]
+            if PROF_DEFAULT_STR == field[:LEN_DEF_STR]:
+                vals[field[LEN_DEF_STR:]] = vals[field]
+                # prefixed fields must be removed from dict
+                # because they are in profile not in product
+                vals.pop(field)
+        return vals
 
     @api.onchange('profile_id')
     def _onchange_from_profile(self):
         """ Update product fields with product.profile corresponding fields """
-        for field, value in self._get_profile_data(self.profile_id.id).items():
-            try:
-                if PROF_DEFAULT_STR == field[:LEN_DEF_STR]:
-                    self[field[LEN_DEF_STR:]] = (
-                        self.profile_id[field])
-                else:
-                    self[field] = self.profile_id[field]
-            except ValueError as e:
-                raise UserError(format_except_message(e, field, self))
-            except Exception as e:
-                raise UserError("%s" % e)
+        if self.profile_id:
```

Comment:
Yes, and they're limiting the practical application of this over a multi-recordset. They are only thinking this as a UI helper, but this function, making it multi-record aware, can be used in any automation. Not making that, you are forcing user to call the method record by record. The "only" problem is if you have to return a domain or a warning, that is not appropriated in a multi-record environment.